compiler/rustc_span/src/hygiene.rs

```diff
@@ -376,6 +383,9 @@ impl HygieneData {
     }
 
     fn is_descendant_of(&self, mut expn_id: ExpnId, ancestor: ExpnId) -> bool {
+        if expn_id.krate != ancestor.krate {
+            return false;
+        }
         while expn_id != ancestor {
```

Comment:
Perhaps there is another optimization possible in this loop since the compiler doesn't know that `self.expn_data(expn_id).parent` has the same `krate`...